br/pkg/lightning/backend/tidb/tidb.go

```diff
@@ -284,22 +287,27 @@ func (enc *tidbEncoder) Encode(logger log.Logger, row []types.Datum, _ int64, co
 	cols := enc.tbl.Cols()
 
 	if len(enc.columnIdx) == 0 {
-		columnCount := 0
+		columnMaxIdx := -1
 		columnIdx := make([]int, len(columnPermutation))
+		for i := 0; i < len(columnPermutation); i++ {
+			columnIdx[i] = -1
+		}
 		for i, idx := range columnPermutation {
 			if idx >= 0 {
 				columnIdx[idx] = i
-				columnCount++
+				if idx > columnMaxIdx {
+					columnMaxIdx = idx
+				}
 			}
 		}
 		enc.columnIdx = columnIdx
-		enc.columnCnt = columnCount
+		enc.columnCnt = columnMaxIdx + 1
 	}
 
 	// TODO: since the column count doesn't exactly reflect the real column names, we only check the upper bound currently.
 	// See: tests/generated_columns/data/gencol.various_types.0.sql this sql has no columns, so encodeLoop will fill the
 	// column permutation with default, thus enc.columnCnt > len(row).
-	if len(row) > enc.columnCnt {
+	if len(row) < enc.columnCnt {
```

Comment:
Could you please explain why `len(row)` should be greater than or equal to `columnCnt`(which is actual `columenMaxIdx + 1`)?

Suppose a table has 5 columns, namely `a`, `b`, `c`, `d`, `e`. We have a SQL statement in a source file like `insert into t(a, d, c) values(xxx, xxx, xxx)`. Then the `columnPermutation` should be `(0, -1, 2, 1, -1, -1)`, `columnCnt = columnMaxIdx + 1 = 4`.  `len(row) ` is 3 which is smaller than `columnCnt`, so in this scenario, data cannot be imported correctly.

Is my understanding correct？

Reply:
@gozssky In this case, the columnMaxIdx should be 2, so columnCnt = 3.

This check is used to handle there are ignored columns. Giving a source file `insert into t(a, d, c, f) values(xxx, xxx, xxx, xx)`, and ignore field `f` in lightning's config, this is still a valid source file. Then len(row) (4) will bigger than columnCnt (3).

BTW, there may be case that row count is smaller than column count if user manually create table schema has extra column with default value, e.g. `updated timestamp default current_timestamp`, then current lightning will result in an error.

Reply:
Oh, I mistook `i` for `idx` 😂.